src/smt/model_blocker.h

```diff
@@ -0,0 +1,69 @@
+/*********************                                                        */
+/*! \file model_blocker.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Utility for blocking the current model
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__THEORY__MODEL_BLOCKER_H
+#define __CVC4__THEORY__MODEL_BLOCKER_H
+
+#include <vector>
+
+#include "expr/expr.h"
+#include "options/smt_options.h"
+#include "theory/theory_model.h"
+
+namespace CVC4 {
+
+/**
+ * A utility for blocking the current model.
+ */
+class ModelBlocker
+{
+ public:
+  /** get model blocker
+   *
+   * This returns a disjunction of literals ~L1 V ... V ~Ln with the following
+   * properties:
+   * (1) L1 ... Ln hold in the current model (given by argument m),
+   * (2) if mode is set to "literals", L1 ... Ln are literals that occur in
+   * assertions and propositionally entail all non-unit top-level assertions.
+   * (3) if mode is set to "values", L1 ... Ln are literals of the form x=c,
+   * where c is the value of x in the current model.
+   * (4) if exprToBlock is not empty, L1 ... Ln are literals of the form t=c,
+   * where t is an element of exprToBlock and c is its value in the current
+   * model.
+   *
+   * We expect exprToBlock to be non-empty if and only if mode is
```

Comment:
if and only if? 
Looks like we only rely on "only if" but not on "if" (in the cpp file):

```
 Assert(mode == BLOCK_MODELS_VALUES);
    std::vector<Node> blockers;
    // if specific terms were not specified, block all variables of
    // the model
    if (nodesToBlock.empty())
    {
      Trace("model-blocker") << "no specific terms to block recognized" << std::endl;
      std::unordered_set<Node, NodeHashFunction> symbols;
      for (Node n : tlAsserts)
      {
        expr::getSymbols(n, symbols);
      }
      for (Node s : symbols)
      {
        if (s.getType().getKind() != kind::FUNCTION_TYPE)
        {
          Node v = m->getValue(s);
          Node a = nm->mkNode(DISTINCT, s, v);
          blockers.push_back(a);
        }
      }
    }
```

Reply:
The "only if" is here: https://github.com/CVC4/CVC4/pull/3112/files#diff-391e93405e2e7168c5717b5b3ade56cdR48

Reply:
right, but where is the "if"? :)

Reply:
Actually it appears the code may need to change here.  The question is what does (block-model-values ()) mean?

Reply:
Currently, if I understand correctly, it is syntactic sugar for `(model-values (x1 ... xn))` with `x1...xn` being the variables in the formula. This was the original implementation of the `values` option.
However, at least for Alessandro's team, this option was much less useful than the potential option of actually specifying the terms they wanted to block. So maybe we shouldn't allow it. What do you think?

Reply:
Yes, let's disallow it. If it were allowed, it would be syntax sugar for `(assert false)` actually.